src/libraries/System.Threading/tests/AutoResetEventTests.cs

```diff
@@ -8,30 +8,127 @@ namespace System.Threading.Tests
+                Assert.Equal(i > 0, es[i].WaitOne(0));
+                es[i].Set();
+            }
+            Assert.Equal(0, WaitHandle.WaitAny(es, ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.Equal(i > 0, es[i].WaitOne(0));
+                es[i].Set();
+            }
+            Assert.Equal(0, WaitHandle.WaitAny(es));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.Equal(i > 0, es[i].WaitOne(0));
+                es[i].Set();
+            }
+            Assert.True(WaitHandle.WaitAll(es, 0));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.False(es[i].WaitOne(0));
+                es[i].Set();
+            }
+            Assert.True(WaitHandle.WaitAll(es, ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.False(es[i].WaitOne(0));
+                es[i].Set();
+            }
+            Assert.True(WaitHandle.WaitAll(es));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.False(es[i].WaitOne(0));
+            }
+
+            // Multi-wait with indexes 1 and 2 set
+            es[1].Set();
+            es[2].Set();
+            Assert.Equal(1, WaitHandle.WaitAny(es, 0));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.Equal(i == 2, es[i].WaitOne(0));
+            }
+            es[1].Set();
+            es[2].Set();
+            Assert.Equal(1, WaitHandle.WaitAny(es, ThreadTestHelpers.UnexpectedTimeoutMilliseconds));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.Equal(i == 2, es[i].WaitOne(0));
+            }
+            es[1].Set();
+            es[2].Set();
+            Assert.False(WaitHandle.WaitAll(es, 0));
+            Assert.False(WaitHandle.WaitAll(es, ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+            for (int i = 0; i < es.Length; ++i)
+            {
+                Assert.Equal(i == 1 || i == 2, es[i].WaitOne(0));
+            }
+
+            // Multi-wait with all indexes reset
+            Assert.Equal(WaitHandle.WaitTimeout, WaitHandle.WaitAny(es, 0));
+            Assert.Equal(WaitHandle.WaitTimeout, WaitHandle.WaitAny(es, ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+            Assert.False(WaitHandle.WaitAll(es, 0));
+            Assert.False(WaitHandle.WaitAll(es, ThreadTestHelpers.ExpectedTimeoutMilliseconds));
+            for (int i = 0; i < es.Length; ++i)
             {
-                Assert.False(are.WaitOne(0));
-                are.Set();
-                Assert.True(are.WaitOne(0));
-                Assert.False(are.WaitOne(0));
-                are.Set();
-                are.Reset();
-                Assert.False(are.WaitOne(0));
+                Assert.False(es[i].WaitOne(0));
             }
         }
```

Comment:
Doesn't have to be in this PR, but these long tests are a bit unwieldy.  Consider creating individual unit tests for each logically separate section, e.g. each section with its own comment header.